browser/ui/views/speedreader/reader_mode_toolbar_view.cc

```diff
@@ -38,23 +44,71 @@ class Toolbar : public views::WebView {
     // Ignore context menu.
     return true;
   }
+
+  raw_ptr<ReaderModeToolbarView> owner_ = nullptr;
 };
 
 }  // namespace
 
-ReaderModeToolbarView::ReaderModeToolbarView(
-    content::BrowserContext* browser_context) {
+void ReaderModeToolbarView::SetDelegate(Delegate* delegate) {
+  delegate_ = delegate;
+}
+
+ReaderModeToolbarView::ReaderModeToolbarView(Browser* browser) {
+  SetVisible(false);
   SetBackground(views::CreateThemedSolidBackground(kColorToolbar));
-  toolbar_ = std::make_unique<Toolbar>(browser_context);
+
+  if (!BraveBrowser::ShouldUseBraveWebViewRoundedCorners(browser)) {
+    SetBorder(views::CreateThemedSolidSidedBorder(
+        gfx::Insets::TLBR(0, 0, 1, 0), kColorToolbarContentAreaSeparator));
+  }
+
+  toolbar_ = std::make_unique<Toolbar>(this, browser->profile());
   AddChildView(toolbar_.get());
 }
 
 ReaderModeToolbarView::~ReaderModeToolbarView() = default;
 
+void ReaderModeToolbarView::SetVisible(bool visible) {
+  if (visible && !toolbar_contents_) {
+    content::WebContents::CreateParams create_params(
+        toolbar_->GetBrowserContext(), FROM_HERE);
+    toolbar_contents_ = content::WebContents::Create(create_params);
+    toolbar_contents_->SetDelegate(this);
+
+    const GURL toolbar_url(kSpeedreaderPanelURL);
+    content::NavigationController::LoadURLParams params(toolbar_url);
+    toolbar_contents_->GetController().LoadURLWithParams(params);
+
+    toolbar_->SetWebContents(toolbar_contents_.get());
+  }
+  views::View::SetVisible(visible);
+}
+
 content::WebContents* ReaderModeToolbarView::GetWebContentsForTesting() {
   return toolbar_->web_contents();
 }
 
+void ReaderModeToolbarView::SwapToolbarContents(
+    ReaderModeToolbarView* another_toolbar) {
+  if (!another_toolbar) {
+    return;
+  }
+
+  CHECK_NE(this, another_toolbar);
+
+  auto* contents = toolbar_->web_contents();
+  auto* another_contents = another_toolbar->toolbar_->web_contents();
+
+  toolbar_->SetWebContents(nullptr);
+  another_toolbar->toolbar_->SetWebContents(nullptr);
+
+  another_toolbar->toolbar_->SetWebContents(contents);
+  contents->SetDelegate(another_toolbar);
```

Comment:
Whenever I create split view with pages that doesn't have speedreader toolbar, crash happens here. Maybe accessing not yet initialzed web_contents from toolbar?

Reply:
Yes, I found it too. The fix is to check `contents` before swapping, I'm adding a tests I'll add one for this case too.